### PR TITLE
Fix replay to pass proof bug

### PR DESCRIPTION
**Pull request by TheWaWaR**

### CODE REVIEW

File: src/merkle_proof.rs [hunk 1]
```diff
@@ -175,10 +177,16 @@ impl MerkleProof {
...
+                        tree_buf.insert((merge_height, parent_key), (leaf_index, leaf_hash));
                         continue;
                     }
-                    let (node, height) = proof.pop_front().ok_or(Error::CorruptedProof)?;
+                    let (height, node_key, left, right_opt) =
+                        proof.pop_front().ok_or(Error::CorruptedProof)?;
+                    let node = if let Some(right) = right_opt {
+                        merge::<H>(height, &node_key, &left, &right)
+                    } else {
+                        left
```
Comment (xxuejie) [inline src/merkle_proof.rs:188]: I was under the impression that right now, even when one branch(left or right) is none, we are still doing the hashing?

File: src/merkle_proof.rs [hunk 2]
```diff
@@ -153,11 +155,11 @@ impl MerkleProof {
         // rebuild the tree from bottom to top
         while !tree_buf.is_empty() {
             // pop_front from tree_buf, the API is unstable
-            let (&(mut height, key), &(leaf_index, node)) = tree_buf.iter().next().unwrap();
```
Comment (jjyr) [inline src/merkle_proof.rs:156]: After the fix, we should do exactly 256 times hashing, so we do not need the `tree_buf` & height in the proof.

File: src/merkle_proof.rs [hunk 3]
```diff
@@ -12,19 +12,20 @@ type Range = core::ops::Range<usize>;
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub struct MerkleProof {
     leaves_path: Vec<Vec<u8>>,
-    proof: Vec<(H256, u8)>,
+    // (height, sibling_key, left, right), when leaf (right = None)
+    proof: Vec<(u8, H256, H256, Option<H256>)>,
```
Comment (jjyr) [inline src/merkle_proof.rs:16]: The new `proof` is `Vec<H256>`, since we should do exactly 256 times hashing, we do not need to save height in the proof.

**Review by jjyr — CHANGES_REQUESTED** [review]

File: src/merkle_proof.rs (hunk 1 above)
Comment (TheWaWaR) [reply src/merkle_proof.rs]: Yes. Only if left/right both are zero we don't do hash.

**Comment by jjyr** [comment]
https://github.com/nervosnetwork/sparse-merkle-tree/pulls

**Comment by TheWaWaR** [comment]
https://github.com/nervosnetwork/sparse-merkle-tree/pull/1

**Pull request closed without merging**